### PR TITLE
Interpolation of offsets in basis time windows

### PR DESCRIPTION
**Pull request by siuwuncheung**
This PR introduces a new style of offsets by interpolation (`-rostype interpolate`) in the parametric ROM setting. There are in total 4 styles of offsets, which can be specified by `-rostype previous`, `-rostype initial`, `-rostype load` and `-rostype interpolate`, where the option `-rparos` becomes redundant in this PR.  In the main file, the default value of `romOptions.offsetType` is now `usePreviousSolution`. 

In the offline phase, we run 
`srun -n 1 -p pdebug laghos -m data/cube01_hex.mesh -pt 211 -tf 0.1 -offline -romsvds -romos -rostype interpolate -romsrhs -bef 1.0 -rpar 0 -nwin 2 -tw tw2.csv`
`srun -n 1 -p pdebug laghos -m data/cube01_hex.mesh -pt 211 -tf 0.1 -offline -romsvds -romos -rostype interpolate -romsrhs -bef 1.5 -rpar 1 -nwin 2 -tw tw2.csv`
`srun -n 1 -p pdebug laghos -m data/cube01_hex.mesh -pt 211 -tf 0.1 -offline -romsvds -romos -rostype interpolate -romsrhs -bef 0.5 -rpar 2 -nwin 2 -tw tw2.csv`

The snapshots are then merged by 
`./merge -nset 3 -ef 0.9999 -rhs -nwin 2 -tw tw2.csv`

A reference solution is given by the full-order model by running
`srun -n 1 -p pdebug laghos -m data/cube01_hex.mesh -pt 211 -tf 0.1 -bef 0.5 -writesol -visit`

Then in the online phase, we run 
`srun -n 1 -p pdebug laghos -m data/cube01_hex.mesh -pt 211 -tf 0.1 -online -romhr -romos -rostype interpolate -sfacx 1 -sfacv 32 -sface 32 -soldiff -romgs -romsrhs -bef 0.5 -nwin 2 -tw tw2.csv -twp twpTemp.csv`

### CODE REVIEW

File: laghos.cpp
Comment (dylan-copeland) [inline laghos.cpp:296]: Is this here just so regression tests will pass? Does it need to be removed immediately after PR approval?

**Comment by dylan-copeland** [comment]
@siuwuncheung Please pull from rom-dev as there have been some recent changes.

File: laghos.cpp
Comment (siuwuncheung) [reply laghos.cpp:296]: @dylan-copeland Yes it is kept here just to make sure the regression tests are passed, but do not appear anywhere in the main program. This line and `-rparos` options in the regression test scripts will be removed right after the PR approval.

**Comment by siuwuncheung** [comment]
@chldkdtn @dylan-copeland @kevinhkhuynh This PR is ready for review. Thanks.

File: laghos.cpp
Comment (dylan-copeland) [inline laghos.cpp]: Extra white line

File: laghos.cpp
Comment (dylan-copeland) [inline laghos.cpp]: Several lines among 791-798 are quite long and should be shortened (based on the usual style of mfem and laghos).

File: laghos.cpp
Comment (dylan-copeland) [inline laghos.cpp]: It looks like the local variable `rom_paramID` can be removed and replaced by `romOptions.parameterID` which should then have the default value of -1 set in laghos_rom.hpp:73.

File: laghos.cpp
Comment (dylan-copeland) [inline laghos.cpp]: What is the purpose of these calculations involving `rhoFactor` and `blast_energyFactor`?

File: tests/sedov-blast/sedov-blast-parameter-variation.sh
Comment (dylan-copeland) [inline tests/sedov-blast/sedov-blast-parameter-variation.sh:5]: If the default value of `-rostype` is `previous`, this does not need to be specified, right? Anyway, we can keep it like this for clarity.

File: laghos.cpp
Comment (siuwuncheung) [reply laghos.cpp]: They are to compute the distances from online parameters to each offline parameter, which is further used to determine the coefficients with respect to the offline parameters in the interpolation of the offset in the online stage. The calculations may have to be relocated to `laghos_rom.cpp` for a better format.

File: tests/sedov-blast/sedov-blast-parameter-variation.sh
Comment (siuwuncheung) [reply tests/sedov-blast/sedov-blast-parameter-variation.sh:5]: Yes they are not needed in this branch and any future branches. They are added to make sure the regression tests results match with the current rom-dev branch, which use `load` as default, as well as make it more clear.

File: laghos.cpp
Comment (dylan-copeland) [reply laghos.cpp]: @siuwuncheung Thanks for the explanation. Should these distances by normalized by the range of the parameter space? For example, if `rhoFactor` ranges over (0.9, 1.1), should there be a normalization based on the range width 0.2? In any case, could you please document how these distances are used in the interpolation?

File: laghos.cpp
Comment (dylan-copeland) [inline laghos.cpp]: Long line

File: laghos.cpp
Comment (siuwuncheung) [reply laghos.cpp]: @dylan-copeland I have chosen the [inverse distance weighting](https://en.wikipedia.org/wiki/Inverse_distance_weighting) for interpolation. The advantages are that 
1. we can make sure the interpolation gives a convex combination of the data points, i.e. stored offsets, 
2. the normalization is provided automatically by sum of inverse distances, without the need of explicitly specifying the range of the varying parameter,
3. can be used for interpolation on multidimensional inputs, i.e. varying both blast energy and density simultaneously.

It is only a particular choice and may have its drawback though, so feel free to suggest another method if we want to compare. Also, may I ask if you think it is better to leave it the way as it is (in `laghos.cpp`), or to pass `rhoFactor` and `blast_energyFactor` to `ROMBasis` and compute the coefficients in each declaration of new basis (in `laghos_rom.cpp`)?

Comment (dylan-copeland) [reply laghos.cpp]: @siuwuncheung Thanks, please just add a comment about the use of inverse distance weighting for interpolation. I agree that this piece of code should be in `laghos_rom.cpp`. It hurts the readability of the main function. Also, it seems the only inputs are the scalar factors, and the output is `romOptions`. This seems straightforward to put in the `ROM_Basis` constructor.

Comment (siuwuncheung) [reply laghos.cpp]: @dylan-copeland Thanks for your suggestion. I added `rhoFactor` and `blast_energyFactor` to `romOptions`, made `coeff_List` and `paramID_List` private members of `ROMBasis` and relocated the calculation to `ROMBasis`. Hope it makes more sense now.

**Review by dylan-copeland — APPROVED** [review]

**Review by chldkdtn — APPROVED** [review]

**Review by kevinhkhuynh — APPROVED** [review]

**Pull request merged**
